### PR TITLE
c-s: basic CLI arguments parser

### PR DESCRIPTION
**Pull request by muzarski**
cassandra-stress accepts the CLI arguments of the following form:
`./cassandra-stress COMMAND [command_params...] [OPTION [option_params...] ...]`

An example:
`./cassandra-stress read cl=QUORUM n=10000 -schema replication(factor=1) -mode cql3 native -rate threads=10`

In this case the command is `read` and its parameters are `cl=` and `n=`.
Apart from the command, there are three options:
* -schema with `replication` parameter
* -mode with `cql3` and `native` parameters
* -rate with a single `threads=` parameter

Introduced corresponding modules:
* `settings` - Entry for CLI arguments parsing.
* `settings::param` - Contains the parameter abstraction and exposes parameters' parser API to the users (see `settings::command`).
* `settings::command` - Parses command and its parameters

Currently, only some subset of commands is supported. No options are supported yet.

Supported commands:
* help
* read
* write
* counter_write
* counter_read

It means that, going back to the example, with this commit cassandra-stress frontend is able to parse only part of the arguments - specifically: `./cassandra-stress read cl=QUORUM n=10000`.
Note that these are not the only available `read` command parameters.

### CODE REVIEW

**Comment by avelanarius** [comment]
@muzarski Some feedback about the commit message:
- Please set your name+surname on your GitHub profile (as requested by our IT)
- Please add your work email to your GitHub account, so that the commit is properly linked to your GitHub account (you don't have to make your work email a primary email in GitHub account, so this doesn't impact your personal GitHub account negatively in any way)
- I know what those terms mean, but I'd appreciate more clear definition/description about "argument"/"parameter"/"command" (maybe some examples)

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 1]
```diff
@@ -0,0 +1,395 @@
+use crate::settings::param::{ParamsParser, SimpleParamHandle};
+use anyhow::Result;
+
+use super::{Command, CommandParams};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Uncertainty {
+    pub err: f32,
+    pub ngt: u64,
+    pub nlt: u64,
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:10]: Those fields should be named more descriptively

File: src/bin/cql-stress-cassandra-stress/main.rs [hunk 2]
```diff
@@ -1,3 +1,25 @@
...
+use crate::settings::parse_cassandra_stress_args;
+use anyhow::Result;
+use std::env;
+
+use settings::CassandraStressParsingResult;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // Cassandra-stress CLI is case-insensitive.
+    let payload = match parse_cassandra_stress_args(env::args().map(|arg| arg.to_lowercase())) {
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/main.rs:15]: The current design of `parse_cassandra_stress_args` has side effects of printing parse errors to stdout - this could be redesigned to instead return error messages in `Err` and print them here (or in `CassandraStressParsingResult::Special` value). One potential benefit of that are unit tests that check the error messages (but not really sure if they are that useful).

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        None,
+        "Time to run in (in seconds, minutes or hours)",
+        true,
+    );
+
+    parser.group(vec![
+        &err, &ngt, &nlt, &no_warmup, &truncate, &cl, &serial_cl,
+    ]);
+    parser.group(vec![&n, &no_warmup, &truncate, &cl, &serial_cl]);
+    parser.group(vec![&duration, &no_warmup, &truncate, &cl, &serial_cl]);
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:272]: I know it is described in `ParamsGroup`, but a (repeat) comment here would be helpful:
```
/// Usage: read [err<?] [n>?] [n<?] [no-warmup] [truncate=?] [cl=?] [serial-cl=?]
///  OR
/// Usage: read n=? [no-warmup] [truncate=?] [cl=?] [serial-cl=?]
///  OR
/// Usage: read duration=? [no-warmup] [truncate=?] [cl=?] [serial-cl=?]
```

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            println!();
+        }
+        println!();
+
+        for param in self.params.iter() {
+            print!("  ");
+            param.borrow().print_desc();
+        }
+    }
+}
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:189]: A test here or in `simple_param.rs` would be appreciated - it would also show the relationship between param and handle.

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                acc += " ";
+            }
+            acc += next.as_ref();
+            (false, acc)
+        })
+        .1;
+    let replaced = REPLACE_1.replace_all(&s, "$1");
+    let replaced = REPLACE_2.replace_all(&replaced, "$1");
+    SPLIT_REGEX.split(&replaced).map(&str::to_owned).collect()
+}
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/mod.rs:107]: A small test of this function would be appreciated - this would also make it easier to understand what this function does (it took me a while to understand why `,` is in `REPLACE_1`, `REPLACE_2`).

**Comment by avelanarius** [comment]
@muzarski in your last amended commit you forgot/deleted by mistake the title of the commit - it now starts with:
```
cassandra-stress accepts the CLI arguments of the following form:
`./cassandra-stress COMMAND [command_params...] [OPTION [option_params...] ...]`
```

**Review by avelanarius — APPROVED** [review]
LGTM from a structural viewpoint - now I see how the introduced abstractions (like param handles) are really needed due to complexity of parsing c-s arguments and I don't see a much cleaner (less mutated state) solution.

@muzarski "Fixes:" in the PR description and commit are better suited to be written under each review comment. The way it is now, the commit message is a bit polluted due to that.

We'll wait for @piodul (or @Lorak-mmk) to comment more on Rust aspects of the code.

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+                        return Err(e);
+                    }
+                    _ => (),
+                }
+            }
+            anyhow::ensure!(consumed, "Invalid parameter {}", arg);
+        }
+
+        Ok(())
+    }
```
Comment (piodul) [inline src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:149]: What if all arguments of one group are matched, but there are some additional arguments that matched to different groups? Shouldn't we reject that?

File: src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        self.supplied_by_user
+    }
+
+    fn required(&self) -> bool {
+        self.required
+    }
+
+    fn set_parsed(&mut self) {
+        self.parsed = true;
+    }
```
Comment (piodul) [inline src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs:94]: Why do we need this method at all? It's strange that you first parse the parameter but then you make it appear as if it weren't parsed at all (due to the `self.parsed` flag being false).

I think the `Param` interface should have two methods: one for checking whether given parameter matches to given string, and the other that actually parses it. Only the second method would modify the state of the parameter, so it would simplify reasoning about the state of a `Param`.

File: src/bin/cql-stress-cassandra-stress/settings/command/help.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+-errors               : How to handle errors when encountered during stress
+-schema               : Replication settings, compression, compaction, etc.
+-node                 : Nodes to connect to
+-log                  : Where to log progress to, and the interval at which to do it
+-transport            : Custom transport factories
+-port                 : The port to connect to cassandra nodes on
+-sendto               : Specify a stress server to send this command to
+-graph                : Graph recorded metrics
+-tokenrange           : Token range settings
+-cloudconf            : Configuration file of a serverless Scylla cluster to connect to"
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/help.rs:37]: See the next comment.

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            "help" => Ok(Self::Help),
+            "print" => Ok(Self::Print),
+            "version" => Ok(Self::Version),
+            "write" => Ok(Self::Write),
+            "read" => Ok(Self::Read),
+            "mixed" => Ok(Self::Mixed),
+            "counter_write" => Ok(Self::CounterWrite),
+            "counter_read" => Ok(Self::CounterRead),
+            "user" => Ok(Self::User),
+            _ => Err(anyhow::anyhow!("No command specified")),
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:38]: True that, this is not looking good. This is the result of me trying to use the exact same error messages that c-s use, and the fact that this check wasn't in the function at first. Then I delegated this code to the function and forgot about it.

I'll make it so this function returns an `Option<Self>` and then the caller should print the error message (if None). Does that sound good?

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+
+    fn show(&self) -> &'static str {
+        match self {
+            Self::Serial => "SERIAL",
+            Self::LocalSerial => "LOCAL_SERIAL",
+        }
+    }
+}
+
+pub struct CSReadWriteParams {
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:138]: I agree.

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+    pub fn print_settings(&self) {
+        println!("******************** Stress Settings ********************");
+        self.params.print_settings(&self.command);
+        println!();
+    }
+}
+
+pub enum CassandraStressParsingResult {
+    Special,
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs:36]: Special commands are the commands that exit prematurely (no workload generated). These are `help`, `print` (receives some distribution as argument, and pretty prints it) and `version` (self explanatory).

Will change it to more descriptive name + add some comment.

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs (hunk 6 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:149]: And we indeed do reject that. It may not be clear, but `ParamsGroup::satisfied` guarantees that. As stated in the comment:
```
The group is satisfied \iff all of the following requirements are satisfied:
     - number of consumed parameters by the group is equal to the number of CLI arguments
     - every required parameter has been provided by the user
```
The first bullet ensures that there were no parameters that weren't consumed by the group. Now that I look at this comment `the number of CLI arguments` part is not very clear. Obviously, I meant the number of parameters that correspond to the command/option that is currently being parsed.

File: src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs (hunk 7 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs:94]: I agree that this name is not self-explanatory - will change it to something else.

But answering your question - why do we need it?
Assume we declare two parameters:
```
let param1 = parser.simple_param(prefix="param1=", ..., default=None, ...);
// Note that second parameter has some default value.
let param2 = parser.simple_param(prefix="param2=", ..., default=Some("value"), ...);

// Make it so they belong to different groups, i.e. are mutually exclusive.
parser.group(vec![&param1]);
parser.group(vec![&param2]);
```

Now assume that the user provides `param1=something`. This results in the first group being satisfied and their parameters marked with `set_parsed()` (note that param2 is not being marked). Thanks to that, `param2.get()` will return `None` even though it has some default value (meaning, it didn't need to consume any CLI argument). 

So basically what `set_parsed()` does is that it marks the parameters of the group that was satisfied, so the user calling `.get()` method on param handle receives:
- `Some(_)` from the parameters of the satisfied group (if it was provided, or had some default value)
- `None` from the paramteres that doesn't belong to the satisfied group.

Is that clear?

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    // Concat to single string.
+    let s = args
+        .fold((true, String::from("")), |(is_first, mut acc), next| {
+            if !is_first {
+                acc += " ";
+            }
+            acc += next.as_ref();
+            (false, acc)
+        })
+        .1;
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs:121]: Yeah... I felt like I was doing something wrong here. For sure we can use `join`.

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+            .iter()
+            .any(|param| param.borrow().required() && !param.borrow().supplied_by_user())
+        {
+            return false;
+        }
+
+        // The group is satisfied - it means that the parameters of this group
+        // were successfully parsed and will be returned to the user.
+        for param in self.params.iter() {
+            param.borrow_mut().set_parsed();
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:66]: `set_parsed()` explained here: https://github.com/scylladb/cql-stress/pull/40#discussion_r1267851172.

Regarding this comment:
I agree that this function does more than it should. I'll refactor the code so it will check whether the group is satisfied. Then, the parser would do something like 
```
if let Some(satisfied_group) = self.groups.iter().find(|g| g.satisfied(args_size)) {
    satisfied_group.mark_params_as_parsed(); // obviously these method will have some more descriptive name, just wanted to stick with the set_parsed() naming
    return Ok(());
}

Err(())
```

Sounds reasonable?

File: src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs (hunk 7 above)
Comment (piodul) [reply src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs:94]: I see.

Now that I think of it, how are you going to detect, from the outside, which group has been parsed? By checking which options return Some/None? Perhaps `ParamsParser::parse` should return the index of the matched group? In that case I don't think `set_parsed` should be necessary.

File: src/bin/cql-stress-cassandra-stress/settings/param/mod.rs [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+/// For now, it may seem to be unnecessary since we only support `SimpleParam`s.
+/// However, cassandra-stress supports more complex parameters (see help -schema)
+/// which cql-stress should support in the future as well.
+pub trait Param {
+    fn parse(&mut self, arg: &str) -> ParamParseResult;
+    fn supplied_by_user(&self) -> bool;
+    fn required(&self) -> bool;
+    fn set_parsed(&mut self);
+    fn print_usage(&self);
+    fn print_desc(&self);
```
[the remark this answers is not in the document]
Comment (piodul) [reply src/bin/cql-stress-cassandra-stress/settings/param/mod.rs:26]: By the way, I think that this interface has too much responsibility. The core part that needs to be abstracted away is the part about parsing the parameter. Basically, it looks like everything apart from `parse` could be implemented by a concrete wrapper struct over a `Param`.

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (piodul) [reply src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:66]: Sounds better.

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (piodul) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:38]: I don't think we need to be 100% compatible with respect to the error messages that are generated when parsing options fails. Basically, we our output to be compatible with the original implementation to the extent that is is usable with our existing tests - which are focused on running commands with correct options. Therefore, I think we can allow ourselves to be a little clearer here.

As for returning `Option<Self>`, I don't have a strong opinion about it, but I'm not opposing it either.

File: src/bin/cql-stress-cassandra-stress/settings/command/help.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/help.rs]: Fixed - still would like to receive a feedback about new solution, tho.

Introduced:
- static array `SUPPORTED_OPTIONS` in `settings::command::mod.rs', which holds currently supported commands. Not sure if this is the right solution. The best solution would be to iterate over `Command` enum's values, but it would require importing new crate: which provides such mechanism https://crates.io/crates/strum. Not sure it that's worth it, tho.
- `Command::print_supported_commands`
- `Command::print_short_description`
- `Command::print_help`

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs]: Fixed.

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs]: Fixed

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    let mut cmd: &str = "";
+    for (i, arg) in args.iter().enumerate() {
+        let arg = arg.as_ref();
+        if i == 0 {
+            cmd = arg;
+        }
+        if i == 0 || arg.starts_with('-') {
+            anyhow::ensure!(
+                !cl_args.contains_key(arg),
+                "{} is defined multiple times. Each option/command can be specidied at most once.",
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs]: Fixed.

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    Ok((cmd, cl_args))
+}
+
+lazy_static! {
+    // Removes whitespaces before characters: ,=()
+    static ref REPLACE_1: Regex = Regex::new(r"\s+([,=()])").unwrap();
+    // Removes whitespaces after characters: ,=(
+    static ref REPLACE_2: Regex = Regex::new(r"([,=(])\s+").unwrap();
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs]: Fixed

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs]: Fixed, changed to `join`.

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs]: Fixed - changed to more descriptive name `SpecialCommand` and added comment about which commands are special.

**Comment by piodul** [comment]
@muzarski - when you push a new version of the PR, please post a comment with a list of changes since the last version - like this:

> v2:
> - changed this,
> - changed that,
> - changed the other thing as well

File: src/bin/cql-stress-cassandra-stress/settings/param/parser.rs (hunk 6 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/parser.rs]: Fixed.
Got rid of this function. I realised it's not necessary to delegate the parsing to the groups. The parser holds the parameters' references as well so he can pretty much do it on his own. The groups are only now used to check for their "satisfaction".

File: src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs (hunk 7 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs]: > By checking which options return Some/None?
Exactly. The resulting structs hold the `Option`s as well in some cases.
> Perhaps ParamsParser::parse should return the index of the matched group? In that case I don't think set_parsed should be necessary.
That's true, it wouldn't be necessary then. I even thought of it for a moment during implementation. But I am not quite sure how then extraction of the values could look like. Currently, we extract the values via `SimpleParamHandle` API.

File: src/bin/cql-stress-cassandra-stress/settings/param/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/mod.rs]: > By the way, I think that this interface has too much responsibility. The core part that needs to be abstracted away is the part about parsing the parameter. Basically, it looks like everything apart from `parse` could be implemented by a concrete wrapper struct over a `Param`.

That's definitely correct. I recognized this issue when implementing `MultiParam` (will be introduced in next PR). I had to copy some of the boilerplate code e.g. `set_satisfied` or `required` which looked exactly the same for both `SimpleParam` and 'MultiParam`. I'll do it then - delegate some of the logic to some wrapper struct.

**Comment by muzarski** [comment]
v2:
- renamed `CSReadWriteParams` to `ReadWriteParams`
- changed concatenation to single string in `repair_params` from using `fold` to `join`
- changed the names of `REPLACE_1` and `REPLACE_2` to be more descriptive
- got rid of the `parse_groups` function in the parser. The parser tries to parse the arguments on his own (since he holds the references to parameters as well). The groups are now only used to check for their satisfaction.
- Changed the `Param` trait interface. Added descriptive comments. There is still a TODO: delegate everything except `parse` to some wrapper struct.
- Introduced `SUPPORTED_COMMANDS` and added some `Command` functions for help printing.
- Introduced some additional comments throughout the code.

**Comment by muzarski** [comment]
v3:
- refactored the command parsing module
- introduced `CommandParser` and `build_command_parser` function which builds the parser for supported commands.

**Comment by muzarski** [comment]
@piodul - as discussed yesterday, let's postpone the decision on what approach to choose when parsing parameters until we introduce more complex parameters. It should be more clear then, and easier to understand how to design the solution.

Please, take a one more look at the changes and provide some feedback.

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+        }
+    }
+
+    fn show(&self) -> &'static str {
+        match self {
+            Self::Never => "NEVER",
+            Self::Once => "ONCE",
+            Self::Always => "ALWAYS",
+        }
+    }
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:65]: > Another question: why `parse` matches on lowercase, but `show` generates an uppercase string?

Since c-s CLI is case insensivite, I convert everything to lowercase before parsing (and so in the parsing method I assume that everything is in lowercase). But from what I've seen, the enum types such as `Truncate` or `ConsistencyLevel` are usually passed in upper case, that's why i print them like this.

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+pub fn build_command_parser() -> CommandParser {
+    let mut parser = CommandParser::new();
+    parser.register_command("read", Command::Read);
+    parser.register_command("write", Command::Write);
+    parser.register_command("counter_write", Command::CounterWrite);
+    parser.register_command("counter_read", Command::CounterRead);
+    parser.register_command("help", Command::Help);
+    parser
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:145]: I agree, I even saw that there is a `strum` crate I could use here. But I wasn't sure if it's worth importing this crate just for this functionality. Should have asked you about it.

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        let params = command.parse_params(cl_args)?;
+
+        Ok((command, params))
+    }
+
+    pub fn print_help(&self) {
+        println!("---Commands---");
+        for (_, cmd) in self.commands.iter() {
+            cmd.print_short_description();
+        }
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:134]: I think iterating over possible enum values (with `strum`) solves this issue anyway (mentioned in the following comments).

File: src/bin/cql-stress-cassandra-stress/settings/command/mod.rs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+        cl_args: &mut ParsePayload,
+    ) -> Result<(Command, CommandParams)> {
+        let command = match self.commands.get(command_str) {
+            None => return Err(anyhow::anyhow!("Invalid command specified")),
+            Some(cmd) => cmd.clone(),
+        };
+
+        let params = command.parse_params(cl_args)?;
+
+        Ok((command, params))
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/mod.rs:127]: True. TBH, the map was only used to generate the help message. But we can easily generate the help message using `strum` which I'll import.

**Comment by muzarski** [comment]
v4:
- Imported `strum` and `strum_macros` crates
- Refactored command parsing module
- `Command`, `Truncate`, `ConsistencyLevel` and `SerialConsistencyLevel` parsing and displaying make use of `strum` functionalities.

**Comment by muzarski** [comment]
v5:
- Rebased on top of https://github.com/scylladb/cql-stress/commit/0d2c190ba267e5b7fd4154246b4c163bfb40385e

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 21]
```diff
@@ -0,0 +1,414 @@
...
+pub enum SerialConsistencyLevel {
+    Serial,
+    LocalSerial,
+}
+
+impl SerialConsistencyLevel {
+    fn parse(serial_cl: &str) -> Self {
+        match Self::from_str(serial_cl) {
+            Ok(serial_cl) => serial_cl,
+            Err(_) => panic!("Invalid serial consistency level: {serial_cl}"),
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:108]: This function is called after `ParamParser::parse()` function which makes sure that `serial-cl=` parameter's value matches regexp `^(serial|local_serial)$`. If the pattern doesn't match, we return an error (and exit gracefully). So at this point (if parser's API is used correctly by the developer) we can be sure that `SerialConsistencyLevel::parse` won't panic.

But I can still change it to return Error if necessary. What do you think?

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 22]
```diff
@@ -0,0 +1,414 @@
...
+        truncate,
+        consistency_level,
+        serial_consistency_level,
+        operation_count,
+        duration,
+    }
+}
+
+pub fn parse_read_write_params(cmd: &Command, payload: &mut ParsePayload) -> Result<CommandParams> {
+    println!("{}", cmd.show());
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:319]: I had to leave it by mistake. Will remove it.

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (piodul) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:108]: > This function is called after `ParamParser::parse()` function which makes sure that `serial-cl=` parameter's value matches regexp `^(serial|local_serial)$`. If the value doesn't match the pattern, we return an error (and exit gracefully). So at this point (if parser's API is used correctly by the developer) we can be sure that `SerialConsistencyLevel::parse` won't panic.

The problem is that, if one looks at these `parse` functions in isolation, they may not have this context and wonder why don't handle the errors correctly and whether those functions aren't misused somewhere. Panics should be avoided in production code, but can be used sparingly as long as it is clear why they are there and we won't stumble upon them accidentally.

Besides, I'm a fan of the ["parse, don't validate" mantra](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) myself. It's unfortunate that we separate validation from parsing, but let's not change that now.

> But I can still change it to return Error if necessary. What do you think?

Yes, I think you should. It should be up to the caller to perform `unwrap()` or panic explicitly in some other way, in a place where it is more obvious that panicking on error is OK.

**Comment by muzarski** [comment]
v6:
- Removed unnecessary print in `parse_read_write_params`
- Changed the parsing logic of some functions in `read_write.rs` to return `Err` instead of `panic` when parsing fails.

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs]: > Besides, I'm a fan of the ["parse, don't validate" mantra](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) myself. It's unfortunate that we separate validation from parsing, but let's not change that now.

Thank you for providing the link to the blog. I'll definitely review it so it helps me deciding on what needs to be changed later on.

File: src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs [hunk 23]
```diff
@@ -0,0 +1,406 @@
...
+    Two,
+    Three,
+    LocalOne,
+    Serial,
+    LocalSerial,
+}
+
+impl ConsistencyLevel {
+    fn parse(cl: &str) -> Result<Self> {
+        Self::from_str(cl).context(format!("Invalid consistency level: {}", cl))
```
Comment (piodul) [inline src/bin/cql-stress-cassandra-stress/settings/command/read_write.rs:82]: The `format!` here always allocates a string (because of `format!`). It's better to use `with_context` if you need to perform expensive computation on error:

```rust
Self::from_str(cl).with_context(|| format!("Invalid consistency level: {}", cl))
```

**Comment by muzarski** [comment]
v7:
- Replace `.context` with `.with_context`, so the string allocation is performed lazily once the error actually occurs.

**Review by piodul — APPROVED** [review]
